From: Paul Moses <p@1g4.org>
Subject: [PATCH] gparam: match fundamental type callback signatures

GType calls the fundamental class and instance initializers through
GClassInitFunc and GInstanceInitFunc. Give the private functions those
signatures and cast their arguments inside the functions. Remove the
empty base initializer and finalizer instead of registering cast pointers.

Exercise a parameter subtype without an instance initializer to check
inherited initialization, default values and floating-reference ownership.
Public types and callback signatures remain unchanged.

diff --git a/gobject/gparam.c b/gobject/gparam.c
--- a/gobject/gparam.c
+++ b/gobject/gparam.c
@@ -58,12 +58,10 @@
 #define PSPEC_APPLIES_TO_VALUE(pspec, value)	(G_TYPE_CHECK_VALUE_TYPE ((value), G_PARAM_SPEC_VALUE_TYPE (pspec)))
 
 /* --- prototypes --- */
-static void	g_param_spec_class_base_init	 (GParamSpecClass	*class);
-static void	g_param_spec_class_base_finalize (GParamSpecClass	*class);
-static void	g_param_spec_class_init		 (GParamSpecClass	*class,
-						  gpointer               class_data);
-static void	g_param_spec_init		 (GParamSpec		*pspec,
-						  GParamSpecClass	*class);
+static void g_param_spec_class_init (gpointer g_class,
+                                     gpointer class_data);
+static void g_param_spec_init (GTypeInstance *instance,
+                               gpointer g_class);
 static void	g_param_spec_finalize		 (GParamSpec		*pspec);
 static void	value_param_init		(GValue		*value);
 static void	value_param_free_value		(GValue		*value);
@@ -118,15 +116,15 @@
   const GTypeInfo param_spec_info = {
     sizeof (GParamSpecClass),
 
-    (GBaseInitFunc) g_param_spec_class_base_init,
-    (GBaseFinalizeFunc) g_param_spec_class_base_finalize,
-    (GClassInitFunc) g_param_spec_class_init,
-    (GClassFinalizeFunc) NULL,
-    NULL,	/* class_data */
+    NULL,
+    NULL,
+    g_param_spec_class_init,
+    NULL,
+    NULL, /* class_data */
 
     sizeof (GParamSpec),
-    0,		/* n_preallocs */
-    (GInstanceInitFunc) g_param_spec_init,
+    0, /* n_preallocs */
+    g_param_spec_init,
 
     &param_value_table,
   };
@@ -144,19 +142,11 @@
 }
 
 static void
-g_param_spec_class_base_init (GParamSpecClass *class)
-{
-}
-
-static void
-g_param_spec_class_base_finalize (GParamSpecClass *class)
-{
-}
-
-static void
-g_param_spec_class_init (GParamSpecClass *class,
-			 gpointer         class_data)
-{
+g_param_spec_class_init (gpointer g_class,
+                         gpointer class_data)
+{
+  GParamSpecClass *class = g_class;
+
   class->value_type = G_TYPE_NONE;
   class->finalize = g_param_spec_finalize;
   class->value_set_default = NULL;
@@ -167,9 +157,12 @@
 }
 
 static void
-g_param_spec_init (GParamSpec      *pspec,
-		   GParamSpecClass *class)
-{
+g_param_spec_init (GTypeInstance *instance,
+                   gpointer g_class)
+{
+  GParamSpec *pspec = (GParamSpec *) instance;
+  GParamSpecClass *class = g_class;
+
   pspec->name = NULL;
   pspec->_nick = NULL;
   pspec->_blurb = NULL;
diff --git a/gobject/tests/param.c b/gobject/tests/param.c
--- a/gobject/tests/param.c
+++ b/gobject/tests/param.c
@@ -1654,6 +1654,37 @@
 }
 
 static void
+test_param_spec_no_init (void)
+{
+  const GParamSpecTypeInfo info = {
+    .instance_size = sizeof (GParamSpec),
+    .value_type = G_TYPE_INT,
+  };
+  GType type = g_param_type_register_static ("GParamNoInit", &info);
+  GParamSpec *pspec;
+  const GValue *value;
+
+  pspec = g_param_spec_internal (type, "no-init", "nick", "blurb", G_PARAM_READWRITE);
+
+  g_assert_cmpuint (G_PARAM_SPEC_TYPE (pspec), ==, type);
+  g_assert_cmpuint (pspec->value_type, ==, G_TYPE_INT);
+  g_assert_cmpuint (pspec->owner_type, ==, G_TYPE_INVALID);
+  g_assert_cmpuint (pspec->param_id, ==, 0);
+  g_assert_cmpuint (pspec->flags, ==, G_PARAM_READWRITE);
+  g_assert_cmpstr (g_param_spec_get_name (pspec), ==, "no-init");
+  g_assert_cmpstr (g_param_spec_get_nick (pspec), ==, "nick");
+  g_assert_cmpstr (g_param_spec_get_blurb (pspec), ==, "blurb");
+
+  value = g_param_spec_get_default_value (pspec);
+  g_assert_true (G_VALUE_HOLDS_INT (value));
+  g_assert_cmpint (g_value_get_int (value), ==, 0);
+
+  g_assert_true (g_param_spec_ref_sink (pspec) == pspec);
+  g_assert_cmpuint (pspec->ref_count, ==, 1);
+  g_param_spec_unref (pspec);
+}
+
+static void
 test_param_spec_pool (void)
 {
   GParamSpecPool *pool = g_param_spec_pool_new (FALSE);
@@ -1725,6 +1756,7 @@
   g_test_add_func ("/paramspec/variant", test_param_spec_variant);
   g_test_add_func ("/paramspec/variant/cmp", test_param_spec_variant_cmp);
   g_test_add_func ("/paramspec/custom", test_param_spec_custom);
+  g_test_add_func ("/paramspec/no-init", test_param_spec_no_init);
   g_test_add_func ("/paramspec/pool", test_param_spec_pool);
 
   return g_test_run ();
